src/voxel_annotation/frontend.ts

```diff
@@ -0,0 +1,727 @@
+    for (let x = -1; x <= 1; x++) {
+      if (x === 0 && y === 0 && z === 0) continue;
+      OFFSETS_26_CONNECTED.push([x, y, z]);
+    }
+  }
+}
+
+@registerSharedObjectOwner(VOX_EDIT_BACKEND_RPC_ID)
+export class VoxelEditController extends SharedObject {
+  public undoCount = new WatchableValue<number>(0);
+  public redoCount = new WatchableValue<number>(0);
+
+  constructor(private host: VoxelEditControllerHost) {
+    super();
+    const rpc = this.host.rpc;
+    if (!rpc) {
+      throw new Error(
+        "VoxelEditController: Missing RPC from multiscale chunk manager.",
+      );
+    }
+
+    const sourcesByScale = this.host.primarySource.getSources(
+      this.getIdentitySliceViewSourceOptions(),
+    );
+    const sources = sourcesByScale[0];
+    if (!sources) {
+      throw new Error(
+        "VoxelEditController: Could not retrieve sources from multiscale object.",
+      );
+    }
+
+    const resolutions: VoxelLayerResolution[] = [];
+
+    for (let i = 0; i < sources.length; ++i) {
+      const source = sources[i]!.chunkSource;
+      const rpcId = source.rpcId;
+      if (rpcId == null) {
+        throw new Error(
+          `VoxelEditController: Source at LOD index ${i} has null rpcId during initialization.`,
+        );
+      }
+      resolutions.push({
+        lodIndex: i,
+        transform: Array.from(sources[i]!.chunkToMultiscaleTransform),
+        chunkSize: Array.from(source.spec.chunkDataSize),
+        sourceRpc: rpcId,
+      });
+    }
+
+    this.initializeCounterpart(rpc, { resolutions });
+  }
+
+  private morphologicalConfig = {
+    growthThresholds: [
+      { count: 100, size: 1 },
+      { count: 1000, size: 3 },
+      { count: 10000, size: 5 },
+      { count: 100000, size: 7 },
+    ],
+    maxSize: 9,
+  };
+
+  readonly singleChannelAccess: ChunkChannelAccessParameters = {
+    numChannels: 1,
+    channelSpaceShape: new Uint32Array([]),
+    chunkChannelDimensionIndices: [],
+    chunkChannelCoordinates: new Uint32Array([0]),
+  };
+
+  private getIdentitySliceViewSourceOptions() {
+    const rank = this.host.primarySource.rank as number | undefined;
+    if (!Number.isInteger(rank) || (rank as number) <= 0) {
+      throw new Error("VoxelEditController: Invalid multiscale rank.");
+    }
+    const r = rank as number;
+    const displayRank = r;
+    const multiscaleToViewTransform = new Float32Array(displayRank * r);
+    for (let chunkDim = 0; chunkDim < r; ++chunkDim) {
+      for (let displayDim = 0; displayDim < displayRank; ++displayDim) {
+        multiscaleToViewTransform[displayRank * chunkDim + displayDim] =
+          chunkDim === displayDim ? 1 : 0;
+      }
+    }
+    return {
+      displayRank,
+      multiscaleToViewTransform,
+      modelChannelDimensionIndices: [],
+    } as const;
+  }
+
+  private getSourceForLOD(
+    multiscale: MultiscaleVolumeChunkSource | undefined,
+    lodIndex: number,
+  ): VolumeChunkSource {
+    if (!multiscale)
```

Comment:
I think it is best to use a block statement with if statements that aren't single line